File: 10. Dynamic Programming/01. Climbing Stairs/main.py
# 1. Brute Force
def climb_brute(n):
    if n <= 1:
        return 1
    return climb_brute(n - 1) + climb_brute(n - 2)

# 2. Memoization
def climb_memo(n, memo=None):
    if memo is None:
        memo = {}
    if n <= 1:
        return 1
    if n in memo:
        return memo[n]
    memo[n] = climb_memo(n - 1, memo) + climb_memo(n - 2, memo)
    return memo[n]

# 3. Tabulation
def climb_tab(n):
    if n <= 1:
        return 1
    dp = [0] * (n + 1)
    dp[0], dp[1] = 1, 1
    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return dp[n]

# 4. Space Optimized
def climb_opt(n):
    if n <= 1:
        return 1
    a, b = 1, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b

# Main function
def main():
    n = 5
    print("Brute:", climb_brute(n))
    print("Memo:", climb_memo(n))
    print("Tabulation:", climb_tab(n))
    print("Optimized:", climb_opt(n))

main()

File: 10. Dynamic Programming/01. Climbing Stairs/main.go
package main

import "fmt"

// 1. Brute Force
func climbBrute(n int) int {
	if n <= 1 {
		return 1
	}
	return climbBrute(n-1) + climbBrute(n-2)
}

// 2. Memoization
func climbMemo(n int, memo []int) int {
	if n <= 1 {
		return 1
	}
	if memo[n] != -1 {
		return memo[n]
	}
	memo[n] = climbMemo(n-1, memo) + climbMemo(n-2, memo)
	return memo[n]
}

// 3. Tabulation
func climbTab(n int) int {
	if n <= 1 {
		return 1
	}
	dp := make([]int, n+1)
	dp[0], dp[1] = 1, 1
	for i := 2; i <= n; i++ {
		dp[i] = dp[i-1] + dp[i-2]
	}
	return dp[n]
}

// 4. Space Optimized
func climbOpt(n int) int {
	if n <= 1 {
		return 1
	}
	a, b := 1, 1
	for i := 2; i <= n; i++ {
		a, b = b, a+b
	}
	return b
}

func main() {
	n := 5

	memo := make([]int, n+1)
	for i := range memo {
		memo[i] = -1
	}

	fmt.Println("Brute:", climbBrute(n))
	fmt.Println("Memo:", climbMemo(n, memo))
	fmt.Println("Tabulation:", climbTab(n))
	fmt.Println("Optimized:", climbOpt(n))
}
